messaging/client.go

```diff
@@ -242,235 +210,356 @@ func (c *Client) Publish(m *Message) (uint64, error) {
-			"replicaID": {strconv.FormatUint(replicaID, 10)},
-			"topicID":   {strconv.FormatUint(topicID, 10)},
-		}.Encode()
-		req, _ := http.NewRequest("DELETE", u.String(), nil)
-		resp, err = http.DefaultClient.Do(req)
-		if err != nil {
-			return err
-		}
-		defer func() { _ = resp.Body.Close() }()
-
-		// If a temporary redirect occurs then update the leader and retry.
-		// If a non-204 status is returned then an error occurred.
-		if resp.StatusCode == http.StatusTemporaryRedirect {
-			redirectURL, err := url.Parse(resp.Header.Get("Location"))
-			if err != nil {
-				return fmt.Errorf("bad redirect: %s", resp.Header.Get("Location"))
-			}
-			c.SetLeaderURL(redirectURL)
-			continue
-		} else if resp.StatusCode != http.StatusNoContent {
-			return errors.New(resp.Header.Get("X-Broker-Error"))
+	// Retrieve the parameters under lock.
+	c.mu.Lock()
+	topicID, index, u := c.topicID, c.index, c.url
+	c.mu.Unlock()
+
+	// Send the message to the messages endpoint.
+	u.Path = "/messaging/heartbeat"
+	u.RawQuery = url.Values{
+		"topicID": {strconv.FormatUint(topicID, 10)},
+		"index":   {strconv.FormatUint(index, 10)},
+	}.Encode()
+	resp, err = http.Post(u.String(), "application/octet-stream", nil)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// If the server returns a redirect then it's not the leader.
+	// If it returns a non-200 code then return the error.
+	if resp.StatusCode == http.StatusTemporaryRedirect {
+		return ErrNoLeader
+	} else if resp.StatusCode != http.StatusOK {
+		if errstr := resp.Header.Get("X-Broker-Error"); errstr != "" {
+			return errors.New(errstr)
 		}
-		break
+		return fmt.Errorf("heartbeat error: %d", resp.StatusCode)
 	}
-
 	return nil
 }
 
 // streamer connects to a broker server and streams the replica's messages.
-func (c *Client) streamer(done chan chan struct{}) {
-	for {
-		// Check for the client disconnection.
-		select {
-		case ch := <-done:
-			close(ch)
-			return
-		default:
-		}
+func (c *Conn) streamer(closing <-chan struct{}) {
+	defer c.wg.Done()
 
-		// TODO: Validate that there is at least one broker URL.
+	// Continually connect and retry streaming from server.
+	var req *http.Request
+	var reqlock sync.Mutex
 
-		// Choose a random broker URL.
-		urls := c.URLs()
-		u := *urls[rand.Intn(len(urls))]
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		for {
+			// Check that the connection is not closing.
+			select {
+			case <-closing:
+				return
+			default:
+			}
 
-		// Connect to broker and stream.
-		u.Path = "/messaging/messages"
-		if err := c.streamFromURL(&u, done); err == errDone {
-			return
-		} else if err != nil {
-			c.Logger.Print(err)
+			// Create URL.
+			u := c.URL()
+			u.Path = "/messaging/messages"
+			u.RawQuery = url.Values{
+				"topicID":   {strconv.FormatUint(c.topicID, 10)},
+				"index":     {strconv.FormatUint(c.Index(), 10)},
+				"streaming": {strconv.FormatBool(c.Streaming())},
+			}.Encode()
+
+			// Create request.
+			reqlock.Lock()
```

Comment:
It's so we don't update the current request while it's being canceled by the transport.